sisl/sparse.py

```diff
@@ -237,6 +237,41 @@ def __init_shape(self, arg1, dim=1, dtype=None, nnzpr=20, nnz=None,
         # Denote that this sparsity pattern hasn't been finalized
         self._finalized = False
 
+    @classmethod
+    def zeros_from_sparsity_union(cls, *spmats, dtype=None, dim=None):
+        """Create a SparseCSR with explicit zeros in all places that *spmats have nonzeros.
+
+        Parameters
+        ----------
+            *spmats : SparseCSR
+                SparseCSRs to find the sparsity pattern union of.
+            dtype : dtype, optional
+                Output dtype. If not given, use the result dtype of the spmats.
+            dim : int, optional
+                If given, the returned SparseCSR will have this as dim.
+        """
+        if not all(spmats[0].shape == m.shape for m in spmats):
+            raise ValueError(
+                f"Cannot find sparsity union of differently shaped csrs: "
+                + " & ".join(str(m.shape) for m in spmats)
+            )
+        shape = spmats[0].shape
+        if dim is not None:
+            shape = shape[:2] + (dim,)
+        dtype = dtype or reduce(np.result_type, (m.dtype for m in spmats))
+        out = cls(shape, dtype=dtype, nnzpr=1, nnz=2)
+
+        out_col = [
+            np.unique(np.concatenate([m.col[m.ptr[r]:m.ptr[r] + m.ncol[r]] for m in spmats]))
+            for r in range(shape[0])
+        ]
+        out.ncol = np.array([len(cols) for cols in out_col], dtype=np.int32)
+        out.ptr = np.cumsum([0] + list(out.ncol), dtype=np.int32)
```

Comment:
I'm going to assume you're right about that...

Reply:
just checked, it *is* ;)

```python
import numpy as np
from time import time

for N in [10, 100, 200, 1000, 5000, 10000]:
    a = (np.random.rand(N) * 1000).astype(np.int32)
    t0 = time()
    b = np.cumsum([0] + list(a), dtype=np.int32)
    print(f"cumsum = {time() - t0}")
    t0 = time()
    b = np.insert(np.cumsum(a, dtype=np.int32), 0, 0)
    print(f"insert = {time() - t0}")
```
small matrices don't matter, large vectors dose.

Reply:
For 5000 elements I get this:
cumsum = 0.0012443065643310547
insert = 5.555152893066406e-05

Reply:
Since we go for speed, this is even better (now my favourite approach ;)) **always** faster!
```python
    b = np.empty(a.size + 1, np.int32)
    b[0] = 0
    np.cumsum(a, out=b[1:])
```